### PR TITLE
Added download button for log file

### PR DESCRIPTION
**Pull request by cheld**
Added button to log view to download the log file.

Alternative would be to open a new tab. Currently, we open already a tab for the log view and I didn't want to open yet another tab. On the other hand, this PR will create lot of files in the download folder that have to be removed by the user. (All files start with the same prefix (log-*) so removing is easy.)

### CODE REVIEW

**Comment by cheld** [comment]
Improves https://github.com/kubernetes/dashboard/issues/1828

File: src/app/frontend/logs/component.js [hunk 1]
```diff
@@ -305,6 +305,15 @@ export class LogsController {
     return 'timer_off';
   }
 
+  /**
+   * Return the link to download the log file
+   * @export
+   * @returns {string}
```
Comment (floreks) [inline src/app/frontend/logs/component.js]: s/returns/return

`returns` is not recognized by google closure compiler so it is just skipped.

File: src/app/backend/resource/container/logs.go [hunk 2]
```diff
@@ -117,8 +113,30 @@ func getRawPodLogs(client *client.Clientset, namespace, podID string, logOptions
 	return string(result), nil
 }
 
+// GetPodLogFile
```
Comment (floreks) [inline src/app/backend/resource/container/logs.go]: s/GetPodLogFile/GetLogFile and add doc

File: src/app/backend/resource/container/logs.go [hunk 3]
```diff
@@ -117,8 +113,30 @@ func getRawPodLogs(client *client.Clientset, namespace, podID string, logOptions
...
+		Follow:     false,
+		Previous:   false,
+		Timestamps: false,
+	}
+	return openStream(client, namespace, podID, logOptions)
+}
+
+
+func openStream(client *client.Clientset, namespace, podID string, logOptions *v1.PodLogOptions) (io.ReadCloser, error) {
+	req := client.Core().RESTClient().Get().
```
Comment (floreks) [inline src/app/backend/resource/container/logs.go]: s/Core/CoreV1 - Core is deprecated

File: src/app/backend/resource/container/logs.go [hunk 4]
```diff
@@ -117,8 +113,30 @@ func getRawPodLogs(client *client.Clientset, namespace, podID string, logOptions
...
+		Name(podID).
+		Resource("pods").
+		SubResource("log").
+		VersionedParams(logOptions, scheme.ParameterCodec)
+	return req.Stream()
+}
+
 // Build logs structure for given parameters.
-func ConstructLogs(podID string, rawLogs string, container string, logSelector *logs.Selection) *logs.LogDetails {
+func ConstructLogDetails(podID string, rawLogs string, container string, logSelector *logs.Selection) *logs.LogDetails {
```
Comment (floreks) [inline src/app/backend/resource/container/logs.go]: Should start with method name

File: src/app/backend/resource/container/logs.go [hunk 5]
```diff
@@ -17,6 +17,8 @@ package container
 import (
 	"io/ioutil"
 
+	"io"
```
Comment (floreks) [inline src/app/backend/resource/container/logs.go]: Remove empty line. Core imports should be grouped together.

File: src/app/backend/handler/apihandler.go [hunk 6]
```diff
@@ -20,6 +20,9 @@ import (
 	"strconv"
 	"strings"
```
Comment (floreks) [inline src/app/backend/handler/apihandler.go]: Remove empty line. Core imports should be grouped together.

File: src/app/backend/handler/apihandler.go [hunk 7]
```diff
@@ -2156,14 +2160,39 @@ func (apiHandler *APIHandler) handleLogs(request *restful.Request, response *res
...
+	k8sClient, err := apiHandler.cManager.Client(request)
+	if err != nil {
+		handleInternalError(response, err)
+		return
+	}
+	namespace := request.PathParameter("namespace")
+	podID := request.PathParameter("pod")
+	containerID := request.PathParameter("container")
+
+	filename := fmt.Sprintf("attachment; filename='logs-from-%v-in-%v.txt'", containerID, podID)
```
Comment (floreks) [inline src/app/backend/handler/apihandler.go]: I'm not sure about this code related to download. How about adding some interface that will handle this outside of apihandler or using some library like [grab](https://github.com/cavaliercoder/grab)?

**Comment by floreks** [comment]
Additionally I see that buttons on logs page (next/previous) are not working for me. Test files for generating logs are generating wrong sizes (500mb is generating 50mb, 1GB is generating 100mb). Also while testing with actual 500mb log I could not open logs page until it stopped generating logs.

File: src/app/frontend/logs/component.js (hunk 1 above)
Comment (cheld) [inline src/app/frontend/logs/component.js]: This is a common issue. I will try to fix all locations in a subsequent PR

File: src/app/backend/resource/container/logs.go (hunk 3 above)
Comment (cheld) [inline src/app/backend/resource/container/logs.go]: Found 3 additional places. Will fix in subsequent PR

File: src/app/backend/handler/apihandler.go (hunk 7 above)
Comment (cheld) [inline src/app/backend/handler/apihandler.go]: Yes, I was not happy with the implementation, too. Currently, apihandler is (mostly) dispatching. This method adds some logic and bloats apihandler even more.

On the other hand, adding the http response object as an input parameter of the business logic is not nice, either. 

Maybe there is another option: what about creating a generic download handler and put it to the handler package?

**Comment by cheld** [comment]
> Also while testing with actual 500mb log I could not open logs page until it stopped generating logs.

I could not reproduce this one.

**Comment by cheld** [comment]
PTAL

**Comment by Globegitter** [comment]
Anything holding up this PR? An addition we would really like to see in the dashboard.

**Comment by floreks** [comment]
Probably not. I'll check it tomorrow. For sure it will be available in next release.

**Comment by cheld** [comment]
I guess @floreks knows I am on vacation :-) I can handle review comments end of week.

File: src/app/backend/resource/container/logs.go [hunk 8]
```diff
@@ -117,8 +112,31 @@ func getRawPodLogs(client kubernetes.Interface, namespace, podID string, logOpti
...
+		Previous:   false,
+		Timestamps: false,
+	}
+	filename := fmt.Sprintf("logs-from-%v-in-%v.txt", container, podID)
+	logStream, err := openStream(client, namespace, podID, logOptions)
+	return filename, logStream, err
+}
+
+func openStream(client kubernetes.Interface, namespace, podID string, logOptions *v1.PodLogOptions) (io.ReadCloser, error) {
+	req := client.CoreV1().RESTClient().Get().
```
Comment (floreks) [inline src/app/backend/resource/container/logs.go]: Any reason to create this temp value? Can't `Stream()` be chained with previous line?

**Comment by Globegitter** [comment]
@cheld any chance for you to look at this PR again? :)

**Comment by floreks** [comment]
@Globegitter are you using our HEAD relases?

**Comment by cheld** [comment]
yes, soon :)

File: src/app/backend/resource/container/logs.go (hunk 8 above)
Comment (cheld) [reply src/app/backend/resource/container/logs.go]: chained it

**Comment by cheld** [comment]
PTAL

**Comment by cheld** [comment]
@floreks can we merge?

**Comment by Globegitter** [comment]
@floreks not at the moment, but might consider depending on the ETA for the new release :)

**Comment by floreks** [comment]
ETA for 1.7 is somewhere next week. We'll know more after we test everything.

**Pull request merged**